docker-compose.yaml

```diff
@@ -3,61 +3,95 @@ version: '3.7'
 services:
 
   caddy:
-    container_name: caddy
-    image: abiosoft/caddy:1.0.3-no-stats
+    image: lucaslorentz/caddy-docker-proxy:latest
+    restart: always
     ports:
       - 80:80
       - 443:443
-    network_mode: host
```

Comment:
This isn't due to use a self signed certificate. This is due to the fact that the IPv6 support in Docker is bad. I already encountered this issue and to solve it I had to use this docker image: https://github.com/robbertkl/docker-ipv6nat. This docker image allows to pass the correct IPv6 address to the Docker container.

I will include it in this PR because even if the Linux server doesn't have public IPv6 address it won't interfere with anything because docker-ipv6nat only works with IPv6.

Proof that it isn't about having a self signed certificate:
(First request in IPv6 without docker-ipv6nat and second request in IPv4)

```
caddy_1          | 2020/03/04 17:18:36 http: TLS handshake error from 172.18.0.1:42496: remote error: tls: bad certificate
caddy_1          | 2020/03/04 17:20:57 http: TLS handshake error from 87.64.x.x:63301: remote error: tls: bad certificate
```

Reply:
Sorry, my message was not clear: the thing I don't understand is how caddy is getting the user IPv4, since it is in a docker network:
* that's the reason of ```network_mode: host```
* and because of ```network_mode: host``` here, there is configuration is way more complicated for filtron / searx / morty.

Thank you for the IPv6 information: I've tried to configured a IPv6 without success, now I understand why (also partly because kimsufi IPv6 provide only one IPv6).
As I understand, docker-ipv6nat, caddy may see different (IPv4, source port) for one IPv6 ?

Reply:
The "proxy" from the external network to the internal network of the container doesn't work in IPv6: https://github.com/moby/moby/issues/17666.
That's why you are seeing the IP of the "load balancer": `172.19.0.1` or `172.18.0.1` (depends on how the network is created).
This proxy is called "userland proxy" and you can learn more about it here: https://windsock.io/the-docker-proxy/

Using `network_mode: host` is considered as a very bad practice because it disables the initial network isolation and publish every single port of the container to the "internet". This "option" should be used as a last resort if there is no **other** solution available.

Reply:
Thank you for the link, way more clear for me.

Just to sum up:
* currently, caddy can open any ports, but has no capabilities except NET_BIND_SERVICE and DAC_OVERRIDE (sadly with the root user inside the container, it is not mandatory, another bad point for the abiosoft image).
* to remove this ```--network host```, the full solution is to include docker-ipv6nat with requires the privileged option so to be more or less root on the host ?

So are we moving from some trust to caddy to a blind trust to docker-ipv6nat ?

(without docker-ipv6nat, filtron can't work).

Reply:
The reason why I mainly want to switch to `docker-ipv6nat` is that hardcoded IP address are considered as bad practice because:
- The network `10.10.10.0` could already exist and network collision is already impacting Docker: https://stackoverflow.com/questions/50514275/docker-bridge-conflicts-with-host-network, https://www.lullabot.com/articles/fixing-docker-and-vpn-ip-address-conflicts and more articles on your preferred search engine.
Leaving Docker creating itself the network is the way to go to avoid any collision.
- In a Docker mindset everything should be "temporary" and not tided to real IP addresses that's why it's better to delegate the work to a DNS server

About the security for docker-ipv6nat, you aren't required to give all the permissions. Like explained here: https://github.com/robbertkl/docker-ipv6nat#docker-container *all what's needed* is to give the permissions `NET_RAW`, `NET_ADMIN` and `SYS_MODULE`.
`docker-ipv6nat` has way less attack surface than `caddy` because it doesn't listen to the external world whereas Caddy does and it's *just* a bunch of `iptables` commands.

Also, I wanted to correct some of my previous thoughts. The Searx container isn't in the same network as the Linux server (`network_mode: host`) so any attacks like `doing requests from the Searx instance to any host` (like an internal host) is incorrect but still correct for caddy if one day there is a flaw in it.

Side note:
> (without docker-ipv6nat, filtron can't work).
In IPv4 only environment it does work but not properly in dual stack (IPv4 & IPv6) environment.

Reply:
> In IPv4 only environment it does work but not properly in dual stack (IPv4 & IPv6) environment.

Yes and the problem is the silent fail: everything works except the bot protection when the bots uses an IPv6. (actually I'm not about the filtron behaviour without docker-ipv6nat).

>  [...]but still correct for caddy if one day there is a flaw in it.

yes (if caddy is docker image from scratch, it limits the attack surface)

> hardcoded IP address are considered as bad practice
 
Here a docker-compose.yml without hardcoded IP: https://gist.github.com/dalf/28ffe27e8675553928fcbfbcd5098179
I know there is still the ```--network host``` and the environment variables copy.

Comparaison about the permissions:
|  | docker-ipv6nat<br>(this PR) | lucaslorentz/caddy-docker-proxy<br>(this PR)<br> | caddy<br>(master branch) | Description |
|-------------------------|-----------------------------|--------------------------------------------------|--------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| --privileged | **yes** | no | no | It also lifts all the limitations enforced by the device cgroup controller. In other words, the container can then do almost everything that the host can do. |
| --network host | yes | no | yes | That container’s network stack is not isolated from the Docker host (the container shares the host’s networking namespace), and the container does not get its own IP-address allocated |
| -v /var/run/docker.sock | **yes** | **yes** | no | Full access to docker |
| capabilities |  | NET_BIND_SERVICE | NET_BIND_SERVICE | Bind a socket to Internet domain privileged ports (port numbers less than 1024). |
|  |  | DAC_OVERRIDE | DAC_OVERRIDE | Bypass file read, write, and execute permission checks.  (DAC is an abbreviation of "discretionary access control".) |
|  | **NET_RAW** |  |  | * Use RAW and PACKET sockets;<br>* bind to any address for transparent proxying. |
|  | **NET_ADMIN** |  |  | Perform various network-related operations:<br>* interface configuration;<br>* administration of IP firewall, masquerading, and accounting;<br>* modify routing tables;<br>* bind to any address for transparent proxying;<br>* set type-of-service (TOS)<br>* clear driver statistics;<br>* set promiscuous mode;<br>* enabling multicasting;<br>* use setsockopt(2) to set the following socket options: SO_DEBUG, SO_MARK, SO_PRIORITY (for a priority outside the range 0 to 6), SO_RCVBUFFORCE, and SO_SNDBUFFORCE. |
|  | **SYS_MODULE** |  |  | Load and unload kernel modules (see init_module(2) and delete_module(2)) |

I understand that ```--network host``` and listen on 127.0.0.1 breaks the docker philosophy but I wish to keep the number of permissions minimals.

Actually, if docker-ipv6nat would be part of the docker project, there would the defacto solution. And yes even if it is not the case, there is far less code to audit in docker-ipv6nat than in caddy. 

But still it is weird for me to add more permissions to remove permissions.
I guess there is no real other solution ? (except hack like hard coded iptables rules).